lib/chef/knife/ssh.rb

```diff
@@ -332,7 +332,9 @@ def ssh_command(command, subsession = nil)
         command.force_encoding("binary") if command.respond_to?(:force_encoding)
         subsession.open_channel do |chan|
           if config[:on_error] && exit_status != 0
-            chan.close()
+            chan.exec "true" do |ch, success|
+              raise ArgumentError, "Cannot execute #{command}" unless success
+            end
```

Comment:
firing off a `true` command here and using the success or failure of that command feels very hacky and side-effecty and brittle.